---
title: "Parkinson Meta-Analysis"
author: "Giacomo Antonello"
date: "`r Sys.Date()`"
format: 
  html:
    embed-resources: true
    toc: true
    toc-depth: 2
editor: visual
params:
  data_type: "relative_abundance"
  overwrite_output: "yes"
---

```{r setup}
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE
)

library(mia) # Bioconductor
library(tidyverse) # CRAN
library(magrittr) # CRAN
library(DT) # CRAN
library(parkinsonsMetagenomicData) # remotes::install_github("ASAP-MAC/parkinsonsMetagenomicData")

library(biobakeryUtils) # remotes::install("g-antonello/biobakeryUtils")
```

# Download microbiome data

## PD studies processed with the nextflowMAC pipeline

```{r}

studies_processed <- read_tsv("https://raw.githubusercontent.com/ASAP-MAC/parkinsons_data_search/refs/heads/main/shotgun_samples/processed_datasets.tsv")

studies_processed[,1:5] %>% 
  relocate(`doi.org link`, .after = `Sequence Acc. code`) %>% 
  datatable(
    caption = paste("List of studies processed with the aim to include them in 
                    the meta-analysis so far.",
    "Analysis date:", 
    Sys.Date()))

country_per_study <- studies_processed %>% 
  transmute(
    study_name = case_when(
      grepl("Boktor", `cMD proposed code`) ~ "BoktorJC_2023",
      TRUE ~ `cMD proposed code`
      ),
    Country_ga = trimws(sapply(strsplit(`Country & geography`, ",|\\/"), function(x) x[length(x)]))
  )

```

## See metadata of `parkinsonsMetagenomicData`

```{r}
metadata_datasets_available <- filter(sampleMetadata, 
                                      # keep only studies listed above
                                      study_name %in% studies_processed$`cMD proposed code`, 
                                      # this removes data for which we have no information regarding health-disease status
                                      !is.na(disease))
metadata_datasets_available[, sapply(metadata_datasets_available, function(x) sum(is.na(x))/length(x)) < 0.1] %>% 
  datatable()
```

### write uuids as simple text file

```{r}
writeLines(metadata_datasets_available$uuid, "Data/uuid_available.txt")
```

## Available downloadable data types

```{r}
datatable(parkinsonsMetagenomicData::biobakery_files())
```

## Download parkinsonsMetagenomicData for `r params$data_type`

```{r, eval=params$data_type != "genefamilies_unstratified"}
raw.tse <- returnSamples(data_type = params$data_type,
                                            sample_data = metadata_datasets_available)

# enhance data 
parkinsonsMetaAnalysisData.tse <- pMD_enhance(input.tse = raw.tse, data_type = params$data_type, sampleMetadata.df = metadata_datasets_available)
```

```{r, eval=params$data_type == "relative_abundance"}
parkinsonsMetaAnalysisData.tse <- AddPhyloTree_to_mpa_tse(parkinsonsMetaAnalysisData.tse, CHOCOPhlAn_version = "202403")  
```

```{r, eval=params$data_type == "genefamilies_unstratified"}
# parquet_raw <- arrow::read_parquet(
#   file = "~/Downloads/genefamilies_unstratified_uuid.parquet", col_select = c("gene_family","rpk_abundance", "uuid")
#   )
library(arrow)
file.pq <- open_dataset("~/Downloads/genefamilies_unstratified_uuid.parquet", )

file_filt.pq <- file.pq %>%
  filter(uuid %in% metadata_datasets_available$uuid) %>%
  collect()

# code below crashes due to too much memory used
# data_matrix <- reshape2::acast(file_filt.pq, rpk_abundance ~ gene_family ~ uuid)
data_matrix <- pivot_wider(
  file_filt.pq, 
  names_from = c("gene_family", "uuid"), 
  values_from = "rpk_abundance"
  )
# parquet_raw <- nanoparquet::read_parquet(
#   file = "~/Downloads/genefamilies_unstratified_uuid.parquet", 
#   col_select = c("gene_family","rpk_abundance", "uuid"), 
#   )


parquet_raw <- data.table::as.data.table(parquet_raw)
parquet__subset <- parquet_raw[uuid %in% metadata_] %>%
  filter(uuid %in% metadata_datasets_available$uuid)


raw.tse <- returnSamples(data_type = "genefamilies_unstratified",
                                            sample_data = metadata_datasets_available, 
                         local_files = "~/Downloads/genefamilies_unstratified_uuid.parquet")

tse_intermediate <- pMD_enhance(input.tse = raw.tse, data_type = params$data_type)
```

## Add curated information about the Country of the participants/samples

```{r}
country_per_study.chr <- country_per_study$Country_ga
names(country_per_study.chr) <- country_per_study$study_name

parkinsonsMetaAnalysisData.tse@colData$Country_ga <- country_per_study.chr[parkinsonsMetaAnalysisData.tse@colData$study_name]
```

## Write results

```{r, message=TRUE}
nameSuffix <- paste(parkinsonsMetagenomicData::biobakery_files()$tool[parkinsonsMetagenomicData::biobakery_files()$data_type == params$data_type], params$data_type, sep = "_")
fullOutName <- sprintf("Data/%s.tse", nameSuffix)

if(params$overwrite_output == "yes"){
  unlink(fullOutName, recursive = TRUE)
}

write_TSE_to_dir(parkinsonsMetaAnalysisData.tse, out.dir = fullOutName)

message(sprintf("Data saved in: %s", fullOutName))
```

# Session Info

```{r}
sessionInfo()
```
