---
title: "PD signatures in BugSigDB"
author: "Giacomo Antonello"
date: "`r date()`"
graphics: yes
output:
  rmdformats::readthedown:
    self_contained: true
    code_folding: hide
    toc_depth: 3
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warnings = FALSE,
  message = FALSE)
```

```{r}
# data manipulation packages
library(tidyverse)
library(magrittr)

# table visualization packages
library(reactable)
library(kableExtra)

# data visualization packages
library(pheatmap)
library(ggpubr)
library(plotly)

# bugSigDB packages for literature review
library(bugsigdbr)
library(bugSigSimple)
library(BugSigDBStats)


# to record continent of origin
library(countrycode)
```

# PD studies in `BugSigDB`

We first summarize bacterial signatures found in previous literature. To ensure
reproducibility, we will use bacterial signatures (meaning differentially 
abundant taxa) for Parkinson's Disease found in the BugSigDB resource.

```{r, echo=TRUE}
# define query
condition_of_interest <- "Parkinson disease"
efo_id <- "MONDO:0005180" 
host <- "Homo sapiens"

cbind(c("Condition", "EFO ID", "Host species"), c(condition_of_interest, efo_id, host)) %>% 
  set_colnames(c("Parameter", "Value")) %>% 
  kbl() %>% 
  kable_styling()
```

Table below: Overview of ALL studies related to Parkinson in BugSigDB. The 
table has been sorted to show shotgun studies on top. **Important**: some 
studies appear more than once because curators have reported more than 
one experiment, like comparing multiple groups. This implies potentially 
different cases and controls.

```{r}

bsdb <- importBugSigDB(cache=FALSE)

bsdb_PD_signatures_allMethods <- filter(bsdb, 
       # get studies/signatures on Parkinson alone
       grepl(efo_id, `EFO ID`),
       # restrict search to Homo sapiens
       `Host species` == host
       )

# fix typo of feces,blood

bsdb_PD_signatures_allMethods <- mutate(bsdb_PD_signatures_allMethods,
                                        `Body site` = ifelse(`Body site` == "Feces,Blood", "Blood,Feces", `Body site`))

# create a continent variable
bsdb_PD_signatures_allMethods$Continent <- strsplit(bsdb_PD_signatures_allMethods$`Location of subjects`, ",") %>% 
  lapply(
    function(x){
      tmp <- countrycode(x, origin = "country.name", "continent") 
      tmp[(tmp == "Americas") & (x %in% c("United States of America", "Canada"))] <- "North America"
      tmp[(tmp == "Americas") & !(x %in% c("United States of America", "Canada"))] <- "South America"
      return(tmp)
      }) %>%  
  sapply(paste, collapse = ",")

source(file = "~/Desktop/bugSigSimple_upgrade.R")

# custom, more expanded createStudyTable function
studies_table_clean <- createStudyTable2(bsdb_PD_signatures_allMethods, rows_are = "experiments", includeAlso = c("Sequencing type", "Sequencing platform"))

studies_table_clean %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```

Below the summary function as currently implemented in bugSigSimple

```{r}
createStudyTable(bsdb_PD_signatures_allMethods) %>% 
  reactable(filterable = TRUE)
```

# Some statistics of studies included (For CRN meeting)

```{r}
# NUmber of studies per year

studies_per_year <- bsdb_PD_signatures_allMethods %>% select(PMID, Year) %>% distinct() %>% .$Year %>% table() %>% as.data.frame() %>% arrange(desc(Freq)) %>% set_colnames(c("Year of Publication", "Number of Studies")) 

years_barplot <- ggplot(studies_per_year, aes(x = `Year of Publication`, y = `Number of Studies`)) + 
  geom_col(fill = "#34a270") + 
  geom_text(data = studies_per_year, mapping = aes(x = `Year of Publication`, y = `Number of Studies` + 0.2, label = `Number of Studies`)) + 
  theme_light() +
  scale_y_continuous(breaks = 1:10)

ggsave(filename = "~/Desktop/CRN_MEETING/studies_per_year.png", plot =  years_barplot, height = 4, width = 5, dpi = 300, bg = "transparent")

# Body sites 

bodySites <- bsdb_PD_signatures_allMethods %>% select(PMID, `Body site`) %>% distinct() %>% mutate(`Body site` = ifelse(`Body site` == "Feces,Blood", "Blood,Feces", `Body site`)) %>% .$`Body site` %>% table() %>% as.data.frame() %>% arrange(desc(Freq)) %>% set_colnames(c("Body Site", "Number of Studies")) 

bodySite_barplot <- ggplot(bodySites, aes(x = reorder(`Body Site`, - `Number of Studies`), y = `Number of Studies`)) + 
  geom_col(fill = "#cf2fb3") + 
  geom_text(data = bodySites, mapping = aes(x = reorder(`Body Site`, - `Number of Studies`), y = `Number of Studies` + 1.5, label = `Number of Studies`)) + 
  theme_light() +
  theme(axis.text.x = element_text(angle = -90, hjust = 0), axis.title.x = element_blank())

ggsave(filename = "~/Desktop/CRN_MEETING/BodySites.png", plot = bodySite_barplot, height = 4, width = 5, dpi = 300, bg = "transparent")

# Census methods 

censusMethods <- bsdb_PD_signatures_allMethods %>% select(PMID, `Sequencing type`) %>% distinct() %>% .$`Sequencing type` %>% table() %>% as.data.frame() %>% arrange(desc(Freq)) %>% set_colnames(c("Census Method", "Number of Studies")) 

censusMethods_barplot <- ggplot(censusMethods, aes(x = reorder(`Census Method`, - `Number of Studies`), y = `Number of Studies`)) + 
  geom_col(fill = "#0c8dc3") + 
  geom_text(data = censusMethods, mapping = aes(x = reorder(`Census Method`, - `Number of Studies`), y = `Number of Studies` + 1, label = `Number of Studies`)) + 
  theme_light() +
  theme(axis.title.x = element_blank())

ggsave(filename = "~/Desktop/CRN_MEETING/CensusMethods.png", plot = censusMethods_barplot, height = 4, width = 5, dpi = 300, bg = "transparent")

# Census methods 

location.df <- bsdb_PD_signatures_allMethods %>% select(PMID, `Location of subjects`) %>% distinct() %>% .$`Location of subjects` %>% table() %>% as.data.frame() %>% arrange(desc(Freq)) %>% set_colnames(c("Location of subjects", "Number of Studies")) %>% 
  left_join(bsdb_PD_signatures_allMethods %>% select(`Location of subjects`, Continent), by = "Location of subjects") %>% 
  # for now, remove multi-country studies
  filter(!grepl(",", `Location of subjects`)) %>% 
  distinct() %>% 
  mutate(`Location of subjects` = case_when(
    `Location of subjects` == "United States of America" ~ "USA", 
    TRUE ~ `Location of subjects`
    ))

location_barplot <- ggplot(location.df, aes(x = reorder(`Location of subjects`, - `Number of Studies`), y = `Number of Studies`)) + 
  geom_col(fill = "#e9a64c") + 
  geom_text(data = location.df, mapping = aes(x = reorder(`Location of subjects`, - `Number of Studies`), y = `Number of Studies` + 1, label = `Number of Studies` )) +
  theme_light() +
  facet_wrap(~Continent,scales = "free_x") + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5), axis.title.x = element_blank())
  

ggsave(filename = "~/Desktop/CRN_MEETING/GeographicDistribution.png", plot = location_barplot, height = 5, width = 5, dpi = 300, bg = "transparent")

ggarrange(
  years_barplot, censusMethods_barplot,
  bodySite_barplot, location_barplot,
  ncol = 2
  )

```

# Significant signatures across studies (Binomial test p-value < 0.05)

The function `bugSigSimple::createTaxonTable` generates a list of microbes that 
are most frequently reported in the papers tested. Next, for each microbe, 
it counts how many times it was reported to be enriched or depleted across 
studies. 

To test if a bug is significantly more often reported as increased or decreased,
a binomial test is run, assuming a 0.5 probability for the null hypothesis. The
more disproportionate two groups, the more significant the test is. 

## All studies

```{r}
tmp <- createTaxonTable(bsdb_PD_signatures_allMethods, Inf) %>% 
  filter(`Binomial Test pval` < 0.05) %>% 
  kbl(caption = "Number of significant signatures as found in BugSigDB. Significance is defined as binomial test p-value less than 0.05. Signautres were extracted using bugSigSimple::createTaxonTable") %>% 
  kable_styling()
```

## Studies split by `Sequencing type` 

```{r}

signautres_per_body_site <- bsdb_PD_signatures_allMethods %>% 
  mutate(`Sequencing type` = ifelse(`Sequencing type` == "Feces,Blood", "Blood,Feces", `Sequencing type`)) %>% 
split.data.frame(f = .$`Sequencing type`) %>% 
  # in the list of each body site, report all possible signatures
  lapply(function(x) tryCatch(createTaxonTable(x, n = Inf), error = function(e) return(NULL))) %>% 
  .[!sapply(., is.null)] %>%
  lapply(dplyr::filter, `Binomial Test pval` < 0.05)

cbind(names(signautres_per_body_site), sapply(signautres_per_body_site, nrow)) %>% 
  set_colnames(c("Sequencing type", "Significantly enriched Signatures")) %>% 
  set_rownames(NULL) %>% 
  kbl(caption = "Number of significant signatures as found in BugSigDB subdividing studies by sequencing type, or more generally microbial census method (PCR is not sequencing). Significance is defined as binomial test p-value less than 0.05, testing for each  disproportions. Signautres were extracted using bugSigSimple::createTaxonTable") %>% 
  kable_styling()

```

## Studies spliy by `Body site`

```{r}

signautres_per_body_site <- bsdb_PD_signatures_allMethods %>% 
  mutate(`Body site` = ifelse(`Body site` == "Feces,Blood", "Blood,Feces", `Body site`)) %>% 
split.data.frame(f = .$`Body site`) %>% 
  # in the list of each body site, report all possible signatures
  lapply(function(x) tryCatch(createTaxonTable(x, n = Inf), error = function(e) return(NULL))) %>% 
  .[!sapply(., is.null)] %>%
  lapply(dplyr::filter, `Binomial Test pval` < 0.05)

cbind(names(signautres_per_body_site), sapply(signautres_per_body_site, nrow)) %>% 
  set_colnames(c("Body Site", "Significantly enriched Signatures")) %>% 
  set_rownames(NULL) %>% 
  kbl(caption = "Number of significant signatures as found in BugSigDB. Significance is defined as binomial test p-value less than 0.05. Signautres were extracted using bugSigSimple::createTaxonTable") %>% 
  kable_styling()

```

## Summary of signatures's evidence

  - Without pre-filtering, 26 bugs are differentially abundant in the same 
  direction across studies
  
  - If split by sequencing/census method, only 16S returns significant signatures
  
  - If split by body site, only `Feces` show associations.

We will continue studying all stool signatures only, given that other body sites 
do not show consensus.

# Signatures for each body site in only `Body site = Feces` studies

```{r}

bsdb_PD_signatures_allMethods_onlyFeces <- filter(bsdb_PD_signatures_allMethods, `Body site` == "Feces")
bsdb_PD_signatures_allMethods_onlyFeces.list <- split.data.frame(bsdb_PD_signatures_allMethods_onlyFeces, bsdb_PD_signatures_allMethods_onlyFeces$`Sequencing type`)

some_stats.mxt <- cbind(
  sapply(bsdb_PD_signatures_allMethods_onlyFeces.list, nrow), 
  t(sapply(bsdb_PD_signatures_allMethods_onlyFeces.list, function(x) table(x$`Abundance in Group 1`, useNA = "always")))
  ) %>% 
  set_colnames(c("Total Signatures", "Decreased in PD", "Increased in PD", "NA"))

caption <- paste0("Summary table of BugSigDB signatures reported for Parkinson's disease (EFO/MONDO code ", efo_id, "). Signatures were split by sequencing method in rows. Date of analysis: ", Sys.Date(), ", bugsigdbr package version: ", packageVersion("bugsigdbr"))

kbl(some_stats.mxt, caption = caption) %>% 
  kable_styling()
```

```{r}
createTaxonTable(bsdb_PD_signatures_allMethods_onlyFeces, n = 20) %>% 
  kbl(caption = "List of top 20 signatures found in gut microbiome (Feces) across Parkinson studies regardless of the sequencing technique. The table was generated with bugSigSimple::createTaxonTable") %>% 
  kable_styling()
```

## Shotgun studies only

We are mostly interested in shotgun sequencing studies, as they are more 
comparable across studies.

```{r}
studies_table_clean %>% 
  filter(`Sequencing type` == "WMS") %>% 
  reactable(filterable = TRUE)
```

```{r}

bsdb_PD_signatures_splitMethods <- split.data.frame(bsdb_PD_signatures_allMethods, bsdb_PD_signatures_allMethods$`Sequencing type`)

createTaxonTable(bsdb_PD_signatures_splitMethods$WMS, n= 20) %>% 
  kbl() %>% 
  kable_styling()
```

```{r}

tmp <- bsdb_PD_signatures_splitMethods$WMS %>% 
  mutate(signatureSplit = paste(`MetaPhlAn taxon names`, collapse = ";")) %>% 
  separate_rows(`MetaPhlAn taxon names`, sep = ";") %>% 
  mutate(
    incr_decr = case_when(
      `Abundance in Group 1` == "increased" ~ 1,
      `Abundance in Group 1` == "decreased" ~  - 1,
      TRUE ~ 0
      ),
    FirstAuth = str_split(`Authors list`, pattern = ", ") %>% sapply("[[", 1)) %>% 
  rename(taxaNames = `MetaPhlAn taxon names`) %>% 
  select(taxaNames, FirstAuth, incr_decr) %>% 
  distinct() 

#' Manually remove Clostridiaceae family increase in Rosario D, which 
#' causes a conflict AND is only found increased in PD, in the paper
tmp %<>% 
  filter(!(FirstAuth == "Rosario D" & incr_decr == -1 & grepl("Clostridiaceae", taxaNames)))

tmp_wide <- pivot_wider(tmp, names_from = "FirstAuth", values_from = "incr_decr") %>% 
  column_to_rownames("taxaNames")

rownames(tmp_wide) <- str_split(rownames(tmp_wide), "\\|") %>% sapply(function(x) x[length(x)])

tmp_wide[is.na(tmp_wide)] <- 0

tmp_wide %>% 
pheatmap(legend_breaks = c(-1, 0, 1), legend_labels = c("Decreased", "NA/Unchanged", "Increased"), angle_col = 315)

```


Heatmap of signatures in the same direction in at least 2 studies

```{r}
tmp_wide[rowSums(abs(tmp_wide)) > 2,] %>% 
pheatmap(legend_breaks = c(-1, 0, 1), legend_labels = c("Decreased", "NA/Unchanged", "Increased"), angle_col = 315)
```

As can be seen, most signatures are unique to the study. This is due to the typical
sources of variability in studies, that include, among all:

  1. Data generation steps 
    
    * sample collection
    
    * DNA extraction
    
    * Sequencing platform
  
  2. Data processing steps
    
    * Fastq filtering
    
    * Fastq trimming
    
    * profiling methods
  
  3. Analysis choices
    
    * Prevalence/abundance cutoff
    
    * Data transformation/scaling/normalization
    
    * Model assumptions
    
    * Covariate adjustment

Our meta-analysis will aim to minimize the effect of steps 2 and 3. Adjusting 
for DNA extraction and sequencing method could help minimizing that bias too.

# PubMed querying strategy

## Not all shotgun publications in my list are in PubMed query results

```{r, fig.cap="Venn diagram of shotgun studies included in BugSigDB and Giacomo's lists. NB: Giacomo's list included 4 preprints"}
shotgun_curated_studies_BugSigDB <- studies_table_clean %>% 
  filter(`Sequencing type` == "WMS")

shotgun_curated_studies_GA <- read_tsv("~/Documents/git_repos/parkinsons_data_search/parkinson_shotgun_datasets.tsv")


ggVennDiagram::ggVennDiagram(list("Giacomo" = shotgun_curated_studies_GA$`doi.org link`, "BugSigDB" = shotgun_curated_studies_BugSigDB$DOI)) + 
  scale_x_continuous(expand = c(-1.5,1.5))
```

  * Studies unique to BugSigDB both use Bedarf 2017 dataset to draw conclusions.
  Given that we are interested mostly 
  
# Comments on BugSigDB based on my query

  - Some `Body site`s can be further curated, because there is a "Feces,Blood" 
  study that could be put into the more numerous "Blood,Feces" category. 
  (PMID: 35645785).
  
  - PMID 33293403 curation should be reviewed, I think, for the following reasons:
  
    1. The `Condition` is idiopathic rapid eye movement sleep behavior disorder, 
    not `Parkinson's disease`
    2. `Body site` says "Intestine" but should "Feces"
  
    The curator is "Aishat", no reviewer 
  - The `createStudyTable` is a nice function but could be expanded, maybe to
  include more options, while leaving a default. 
  **I"m working on a GitHub issue or feature request**
  
# IBS signatures in BugSigDB

```{r}
condition_of_interest <- "Irritable Bowel Syndrome"
efo_id <- "EFO:0000555" 
host <- "Homo sapiens"


cbind(c("Condition", "EFO ID", "Host species"), c(condition_of_interest, efo_id, host)) %>% 
  set_colnames(c("Parameter", "Value")) %>% 
  kbl() %>% 
  kable_styling()
```


```{r}
bsdb_IBS_signatures_allMethods <- filter(bsdb, 
       # get studies/signatures on Parkinson alone
       grepl(efo_id, `EFO ID`),
       # restrict search to Homo sapiens
       `Host species` == host
       )
```

```{r}
createStudyTable2(bsdb_IBS_signatures_allMethods) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```


```{r}
createTaxonTable(bsdb_IBS_signatures_allMethods) %>% 
  filter(`Binomial Test pval` < 0.05) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```

# IBD signatures in BugSigDB

```{r}
condition_of_interest <- "Inflammatory Bowel Disease"
efo_id <- "EFO:0003767" 
host <- "Homo sapiens"


cbind(c("Condition", "EFO ID", "Host species"), c(condition_of_interest, efo_id, host)) %>% 
  set_colnames(c("Parameter", "Value")) %>% 
  kbl() %>% 
  kable_styling()
```


```{r}
bsdb_IBD_signatures_allMethods <- filter(bsdb, 
       # get studies/signatures on Parkinson alone
       grepl(efo_id, `EFO ID`),
       # restrict search to Homo sapiens
       `Host species` == host
       )
```

```{r}
createStudyTable2(bsdb_IBD_signatures_allMethods) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```


```{r}
createTaxonTable(bsdb_IBD_signatures_allMethods) %>% 
  filter(`Binomial Test pval` < 0.05) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```

# Constipation signatures in BugSigDB

The following code is run but does not return studies nor signatures. 
That is because, to this date, BugSigDB has not been populated with studies that
have looked directly at **constipation** as `Condition`. 

Feel free to verify, but also a View(bsdb) followed by a general search of 
"constip" did not produce results except in co-occurrence with Autism spectrum
disorder or with Parkinson disease

```{r}
condition_of_interest <- "Constipation"
efo_id <- "NCIT:C37930" 
host <- "Homo sapiens"


cbind(c("Condition", "EFO ID", "Host species"), c(condition_of_interest, efo_id, host)) %>% 
  set_colnames(c("Parameter", "Value")) %>% 
  kbl() %>% 
  kable_styling()
```


```{r}
bsdb_IBD_signatures_allMethods <- filter(bsdb, 
       # get studies/signatures on Parkinson alone
       grepl(efo_id, `EFO ID`),
       # restrict search to Homo sapiens
       `Host species` == host
       )
```

```{r}
createStudyTable2(bsdb_IBD_signatures_allMethods) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```


```{r}
createTaxonTable(bsdb_IBD_signatures_allMethods) %>% 
  filter(`Binomial Test pval` < 0.05) %>% 
  reactable(filterable = TRUE, searchable = TRUE)
```

