---
title: "Parkinson Meta-Analysis"
author: "Giacomo Antonello"
date: "`r Sys.Date()`"
format: 
  html:
    embed-resources: true
    toc: true
    toc-depth: 2
editor: visual
params:
  pMD_data_type: "pathabundance_unstratified"
  pseudocount: 1e-8
  transform: "asin_sqrt"
  prevalThresh: 0.05
---

# Analysis plan

For dataType in {`MetaPhlan`, `HUMAnN UniRef`, `HUMAnN pathways`}:

-   

In datasets with available information of constipation (yes/no and bristol) (meta-) analyse the constipated microbiome in otherwise healthy individuals:

-   Blue Poo paper
-   Natasha's participants
-   

```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE
)

library(mia)
library(magrittr)
library(tidyverse)
library(parkinsonsMetagenomicData)
library(ggtext)

library(ggpubr)
library(meta)

# for table visualizations
library(DT)
library(kableExtra)

library(biobakeryUtils)
theme_set(theme_bw())
```

## Load and further curate data

```{r, cache=TRUE}
input.path <- list.files("Data", pattern = paste0(params$pMD_data_type, ".tse"), full.names = TRUE)
Input.tse <- read_TSE_from_dir(input.path)

# calculate CLR 
Input.tse <- transformAssay(Input.tse, assay.type = "relabundance", method = "clr", pseudocount = as.numeric(params$pseudocount))

# Define A Parkinson_1_0 variable. controls may have other diseases too
Input.tse@colData$Parkinson_1_0 <- factor(ifelse(grepl("Parkinson", Input.tse@colData$disease), 1, 0))

# convert NA in age_group to "unknown"
Input.tse@colData$age_group <- factor(ifelse(is.na(Input.tse@colData$age_group), "Unknown", Input.tse@colData$age_group))

# create number of diseases variables
Input.tse@colData$number_diseases <- ifelse(Input.tse@colData$disease == "Healthy", 0, str_count(Input.tse@colData$disease, ";") + 1)

# Remove all-0 samples
all0_samples <- colnames(assay(Input.tse))[colSums(assay(Input.tse)) == 0]
Input.tse <- Input.tse[, !(Input.tse@colData$uuid %in% all0_samples)]

# assay-wise curations
if(params$transform == "asin_sqrt"){
  assay(Input.tse, params$transform) <- asin(sqrt(assay(Input.tse, "relabundance")))
  }
```

# Aitchison PCA

```{r, eval=FALSE, fig.height=12, fig.width=9}
Input.tse <- addMDS(Input.tse, assay.type = "clr", method = "euclidean", name = "aitchison")

variances_explained <- attr(reducedDim(Input.tse, "aitchison"), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

betadiv_data.df <- cbind.data.frame(as.data.frame(colData(Input.tse)), reducedDim(Input.tse, "aitchison") %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

PCA_aitchison_col_Country <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = Country_ga, fill = Country_ga)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_jco() +
  ggsci::scale_fill_jco() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Country",
       fill = "Country",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_aitchison_col_Study <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = study_name, fill = study_name)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_futurama() +
  ggsci::scale_fill_futurama() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Study",
       fill = "Study",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_aitchison_col_Parkinson <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = Parkinson_1_0)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_bmj() +
  ggsci::scale_fill_bmj() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Parkinson (y/n)",
       fill = "Parkinson (y/n)",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_aitchison_col_age_group <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = age_group)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_bmj() +
  ggsci::scale_fill_bmj() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Age Group",
       fill = "Age Group",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

  
plot_panel_aitchison <- ggarrange(PCA_aitchison_col_Country, PCA_aitchison_col_Study, PCA_aitchison_col_Parkinson, PCA_aitchison_col_age_group)

ggsave(plot = plot_panel_aitchison, filename = "Figures/BetaDiv_Aitchison.svg", height = 14, width = 14)

plot_panel_aitchison
```

```{r PERMANOVA aitchison, eval=FALSE}

Input.tse <- addPERMANOVA(
  Input.tse,
  assay.type = "clr",
  method = "euclidean",
  name = "Aitchison_PERMANOVA",
  formula = x ~ Country_ga + study_name + Parkinson_1_0,
  by = "margin",
  permutations = 2000, 
  parallel = 8
)

biobakeryUtils::PERMANOVA_to_table(Input.tse, name = "Aitchison_PERMANOVA")
write_tsv(biobakeryUtils::PERMANOVA_to_table(Input.tse, name = "Aitchison_PERMANOVA"), file = sprintf("results/03_meta_analysis_Pathway/Aitchison_PERMANOVA.tsv"))
```

# Bray-Curtis PCoA

```{r, eval=FALSE, fig.height=12, fig.width=9}
Input.tse <- addMDS(Input.tse, assay.type = "relabundance", method = "bray")

variances_explained <- attr(reducedDim(Input.tse, "bray"), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

betadiv_data.df <- cbind.data.frame(as.data.frame(colData(Input.tse)), reducedDim(Input.tse, "bray") %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

PCA_bray_col_Country <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = Country_ga, fill = Country_ga)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_jco() +
  ggsci::scale_fill_jco() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Country",
       fill = "Country",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_bray_col_Study <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = study_name, fill = study_name)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_futurama() +
  ggsci::scale_fill_futurama() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Study",
       fill = "Study",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_bray_col_Parkinson <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = Parkinson_1_0)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_bmj() +
  ggsci::scale_fill_bmj() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Parkinson (y/n)",
       fill = "Parkinson (y/n)",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

PCA_bray_col_age_group <- betadiv_data.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = age_group)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  ggsci::scale_color_bmj() +
  ggsci::scale_fill_bmj() +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Age Group",
       fill = "Age Group",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

  
plot_panel_bray <- ggarrange(PCA_bray_col_Country, PCA_bray_col_Study, PCA_bray_col_Parkinson, PCA_bray_col_age_group)

ggsave(plot = plot_panel_bray, filename = "Figures/BetaDiv_bray.svg", height = 14, width = 14)

plot_panel_bray
```

```{r PERMANOVA bray, eval=FALSE}
Input.tse <- addPERMANOVA(
  Input.tse,
  assay.type = "relabundance",
  method = "bray",
  name = "bray_PERMANOVA",
  formula = x ~ Country_ga + study_name + Parkinson_1_0,
  by = "margin",
  permutations = 2000
)

biobakeryUtils::PERMANOVA_to_table(Input.tse, name = "bray_PERMANOVA")
```

# Parkinson meta-analysis with `lm` (via lmFit in limma)

## Unavailability of metadata makes it hard to properly adjust for covariates

```{r}
library(gtsummary)

colData(Input.tse) %>% 
  as.data.frame() %>% 
  select(`Age Gr.` = age_group, Sex = sex, `N. diseases` = number_diseases, Parkinson_1_0) %>% 
   tbl_summary(
    by = Parkinson_1_0, # Optional: stratify by a variable
    statistic = list(all_continuous() ~ "{mean} ({sd})",
                     all_categorical() ~ "{n} ({p}%)"),
    missing_text = "Missing"
  ) %>%
  add_overall() %>% # Add an overall column
  modify_header(label ~ "**Characteristic**") # Customize header
```

## Datasets with available age and sex information

```{r}
colData(Input.tse) %>% 
  as.data.frame() %>% 
  group_by(study_name) %>% 
  
  reframe(
    Tot = n(), 
    avail_sex = sum(!is.na(sex)),
    avail_sex_pct = round(avail_sex/Tot * 100,1),
    avail_age_group = sum(!is.na(age_group)),
    avail_age_group_pct = round(avail_age_group/Tot * 100,1),
    avail_Case_ctrl = sum(!is.na(Parkinson_1_0)),
    avail_Case_ctrl_pct = round(avail_Case_ctrl/Tot * 100,1)
    
  ) %>% 
  datatable()
```

## Perform meta-analysis as suggested in sex meta-analysis vignette

The vignette can be found at [curatedMetagenomicDataAnalyses](https://waldronlab.io/curatedMetagenomicDataAnalyses/articles/Age_metaanalysis_vignette.html) and is designed to perform binary associations. The method is not really advance but it allows for some degree of flexibility.

Analysis choices:

-   Transform: $\arcsin(\sqrt{relative~abundance})$
-   Min prevalence: 5%
-   Min 10 ppl per Case and 10 per control in each dataset
-   Cohen D effect size standardization

## Table of cases and controls in each study

```{r}
cases_ctrls_vs_study_name <- colData(Input.tse) %>% 
  as.data.frame() %>% 
  split.data.frame(.$study_name) %>% 
  lapply(function(x) as.data.frame(table(x$disease, useNA = "always"))) %>% 
  bind_rows(.id = "study_name")
  

cases_ctrls_vs_study_name_wider <- cases_ctrls_vs_study_name %>% 
  pivot_wider(names_from = Var1, values_from = Freq, values_fill = 0)

# create a character vector with study names that can and will be used
# in the meta-analysis
study_names_usable <- cases_ctrls_vs_study_name_wider$study_name[cases_ctrls_vs_study_name_wider$Healthy >= 10 & cases_ctrls_vs_study_name_wider$`Parkinson Disease` >= 10]

# sum cases/controls across studies that will actually be used
colSums_useful_studies <- cases_ctrls_vs_study_name_wider %>% 
  filter(study_name %in% study_names_usable) %>% 
  column_to_rownames("study_name") %>% 
  colSums()


cases_ctrls_vs_study_name_wider %>%
  rbind.data.frame(c("Total", colSums_useful_studies)) %>% 
  
  kbl(caption = "Overview of diseases reported in Parkinson_1_0's disease datasets. Notice that only Wallen has a large documentation of other diseases. We will concentrate only on Parkinson_1_0 and otherwise healthy participants. The last row includes the total number of cases of each disease(s) (column) across all studies") %>% 
  kable_styling() %>% 
  row_spec(which(!(cases_ctrls_vs_study_name_wider$study_name %in% study_names_usable)), color = 
             "gray80")

```

## Filter input data to contain only acceptable studies and taxa with preval \>= \``r params$prevalThresh`%\`

```{r}
Input_filtered.tse <- Input.tse[,Input.tse@colData$study_name %in% study_names_usable]

prevals <- getPrevalence(Input_filtered.tse, assay.type = assayNames(Input_filtered.tse)[1])
  
Input_filtered.tse <- Input_filtered.tse[prevals >= as.numeric(params$prevalThresh),]
```

```{r}
Input_filtered.tse
```

## Perform lm

```{r}
library(limma)

per_dataset_limma_results.list <- Input_filtered.tse %>% 
  splitOn(group = "study_name") %>% 
  
  
  lapply(function(x.tse) {
    x.tse <- x.tse[rowSums(assay(x.tse, params$transform)) > 0,]
    
    modMtx <- model.matrix(~ Parkinson_1_0, data = as.data.frame(colData(x.tse)))  
    
    lmFit_results <- lmFit(assay(x.tse, params$transform), design = modMtx) %>% 
  eBayes() %>% 
  topTable(coef = "Parkinson_1_01", n = Inf, sort.by = "P", confint = TRUE, adjust.method = "BH") %>% 
  transmute(
    Beta = logFC,
    SE =(CI.R - CI.L)/3.92,
    AveExpr, 
    t,
    B,
    P.Value,
    FDR = adj.P.Val,
    NCases = sum(x.tse@colData$Parkinson_1_0 == 1),
    NControls = sum(x.tse@colData$Parkinson_1_0 == 0)) %>% 
  rownames_to_column("FeatureID") %>% 
  as_tibble()      
  return(lmFit_results)
  }
  )


# Add two types of available taxonomies
#per_dataset_limma_results_GTDB.list <- lapply(per_dataset_limma_results.list, function(x) dplyr::right_join(GTDB.df, x %>% mutate(Pathway = gsub("t__", "",FeatureID), FeatureID = NULL), by = "Pathway"))

per_dataset_limma_results.list <- lapply(per_dataset_limma_results.list, function(x) dplyr::right_join(as.data.frame(rowData(Input_filtered.tse)) %>% rownames_to_column("FeatureID"), x, by = "FeatureID"))

```

## Calculate standardized effect sizes and SEs

```{r, meta-analysis functions definition}
source("R/metaAnalysisUtils.R")
source("R/ForestPlots.R")
```

```{r}
per_dataset_limma_results.list <- lapply(per_dataset_limma_results.list, 
                                   function(x.tb){
                                     mutate(x.tb, 
                                            CohenD = computeCohenD(t, NCases, NControls),
                                            CohenSE = computeCohenSE(CohenD, NCases, NControls),
                                            )
                                   }
                                   )

```

```{r}
CohenD_and_SE_allDatasets.tb <- per_dataset_limma_results.list %>% 
  imap(function(x, xnam) 
    transmute(x,
              FeatureID,
         !!paste0("CohenD__", xnam) := CohenD,
         !!paste0("CohenSE__", xnam) := CohenSE
         )
    ) %>% reduce(full_join, by = "FeatureID")

cohenD.mtx <- CohenD_and_SE_allDatasets.tb %>% 
  column_to_rownames("FeatureID") %>% 
  select(contains("CohenD")) %>% 
  t()

cohenSE.mtx <- CohenD_and_SE_allDatasets.tb %>% 
  column_to_rownames("FeatureID") %>% 
  select(contains("CohenSE")) %>% 
  t()
```

```{r}
meta_analysis_results <- sapply(colnames(cohenD.mtx), function(i) runMetaanalysis(cohenD.mtx[,i], cohenSE.mtx[,i]), USE.NAMES = TRUE) %>% 
  # transpose to have features in rows and stats in columns
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("FeatureID") %>% 
  mutate(RE_FDR = p.adjust(RE_P.value, method = "BH"), .after = RE_P.value) %>% 
  arrange(desc(abs(RE_Eff.Size)))

meta_analysis_results <-dplyr::right_join(as.data.frame(rowData(Input_filtered.tse)) %>% rownames_to_column("FeatureID"), meta_analysis_results, by = "FeatureID")

# save files, also valuable intermediates

meta_analysis_to_write <- list(
  "CohenD" = cohenD.mtx %>% as.data.frame() %>% rownames_to_column("FeatureID"),
  "CohenSE" = cohenSE.mtx %>% as.data.frame() %>% rownames_to_column("FeatureID"),
  `Meta-Analysis` = meta_analysis_results
)
```

## write meta-analysis results

```{r}
dir.create("results/03_meta_analysis_Pathway/", showWarnings = FALSE, recursive = TRUE)

openxlsx2::write_xlsx(per_dataset_limma_results.list, file = "results/03_meta_analysis_Pathway/01-limma-regressions.xlsx")
openxlsx2::write_xlsx(meta_analysis_to_write, file = sprintf("results/03_meta_analysis_Pathway/02-meta_analysis_%s_PD_only.xlsx", params$transform))
```

## Results at a glance

```{r}
feats_tot <- nrow(meta_analysis_results)

heterog <- meta_analysis_results %>% 
  filter(RE_het_P.value <= 0.05) %>% 
  nrow()

notSignif <- meta_analysis_results %>% 
  filter(
    RE_het_P.value > 0.05, 
    RE_FDR >= 0.1
    ) %>% 
  nrow()

Positive_assoc <- meta_analysis_results %>% 
  filter(
    RE_het_P.value > 0.05, 
    RE_FDR < 0.1,
    RE_Eff.Size > 0
    ) %>% 
  nrow()

Negative_assoc <- meta_analysis_results %>% 
  filter(
    RE_het_P.value > 0.05, 
    RE_FDR < 0.1,
    RE_Eff.Size < 0
    ) %>% 
  nrow()

cbind(
  c("Tot tested", "N. heterogeneous", "Not Signif. (FDR >= 0.1)", "Signif. Positive assoc.", "Signif. Negative assoc."),
  c(feats_tot, heterog, notSignif, Positive_assoc, Negative_assoc)
) %>% kbl() %>% 
  kable_styling()
```

## Microbiome Forest plot of Standardized eff. sizes

```{r, fig.height=8, fig.width=14}
legend_table <- Input_filtered.tse %>% 
  colData %>% 
  as_tibble %>% 
  group_by(study_name) %>% 
  reframe(N = n(), PD = sum(Parkinson_1_0 == 1), Control = sum(Parkinson_1_0 == 1)) %>% 
  arrange(desc(N)) %>% 
  mutate(study_label = 1:nrow(.), .before = study_name)


ForestPlot <- biobakeryForestPlot(
  meta_input = meta_analysis_results,
  StdEffSizes = cohenD.mtx %>% t,
  studies_Stats.df = legend_table, 
  topN_feat = 20,
  featuresCol = "FeatureID") %>% 
  enrich_y_pathway_label(pathway_rowData = as.data.frame(rowData(Input_filtered.tse)))

forest_plot_with_legend <- ggarrange(ForestPlot, ggtexttable(legend_table, rows = NULL), nrow = 1, widths = c(1, 0.5))

invisible(sapply(c("png", "svg"), function(fmt) ggsave(plot = forest_plot_with_legend, filename = sprintf("results/03_meta_analysis_Pathway/ForestPlot_%s.%s", params$transform, fmt), dpi = 300, height = 8, width = 14)))

forest_plot_with_legend
```

# Session Info

```{r}
sessionInfo()
```
