---
title: "Constipated microbiome in healthy (non-PD) individuals"
author: "Giacomo Antonello"
date: "`r Sys.Date()`"
format: 
  html:
    embed-resources: true
    toc: true
    toc-depth: 2
editor: visual
params:
  pMD_data_type: "pathabundance_unstratified"
  pseudocount: 1e-8
  transform: "asin_sqrt"
  prevalThresh: 0.05
---

```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE
)

library(mia)
library(magrittr)
library(tidyverse)
library(parkinsonsMetagenomicData)
library(ggtext)

library(ggpubr)
library(meta)

# for table visualizations
library(DT)
library(kableExtra)

library(limma)
library(biobakeryUtils)
theme_set(theme_bw())
```

# `AsnicarF_2021`

## Load private metadata and define transit time groups

```{r}
studyNames <- c("WallenZD_2022", "AsnicarF_2021")
bluePoo_vars.chr <- c("bowel_movements_last_7_days", "bristol_stool_score_average_last_3_months", "transit time (days)")
# load separate asnicar metadata
bluePoo_meta_raw.df <- read_tsv("Data/bluePoo/bluePoo_metadata/bluePoo_metadata_ga.tsv") 

# process what we need from this metadata
AsnicarF_meta_select.df <-  bluePoo_meta_raw.df %>% 
  transmute(
    subject_id = sample_alias,
    `transit time (hours)` = `transit time (days)` * 24,
    transitTimeCut = cut(`transit time (hours)`, breaks = c(0, 14, 38, 58,max(`transit time (hours)` + 1e-6, na.rm = TRUE)), right = FALSE),
    transitTimeCutClasses = cut(`transit time (hours)`, breaks = c(0, 14, 38, 58,max(`transit time (hours)` + 1e-6, na.rm = TRUE)), right = FALSE, labels = paste0("C", 1:4)),
    transitTimeCutClassesDef =  case_when(
     transitTimeCutClasses == "C1" ~ "Fast",
     transitTimeCutClasses %in% c("C2","C3") ~ "Normal",
     transitTimeCutClasses == "C4" ~ "Slow"
     ) %>% factor(c("Normal", "Fast", "Slow"))
    )
AsnicarF_pMD.df <- read_tsv("Data/bluePoo/bluePoo_metadata/AsnicarF_2021.tsv")

Asnicar_complete.df <- inner_join(AsnicarF_meta_select.df, AsnicarF_pMD.df, by = "subject_id")

kbl(table(Asnicar_complete.df$transitTimeCutClassesDef, Asnicar_complete.df$transitTimeCutClasses,useNA = "ifany")) %>% 
  kable_styling() %>% 
  column_spec(1, bold = TRUE)
```

## Load AsnicarF_2021 pMD microbiome data (publicly available)

```{r}
# get pMD microbiome profiles 
basic.tse <- returnSamples(Asnicar_complete.df, data_type = "relative_abundance")

Asnicar_complete.tse <- basic.tse %>% 
  biobakeryUtils::pMD_enhance_MetaPhlAn() %>% 
  replace_colData(Asnicar_complete.df %>% as.data.frame() %>% set_rownames(.$uuid))

assay(Asnicar_complete.tse, "asin_hellinger") <- asin(sqrt(assay(Asnicar_complete.tse, "relabundance")))

# get only available covariate variables
Asnicar_complete_no_NA_Constipation.tse <- Asnicar_complete.tse[, !is.na(Asnicar_complete.tse@colData$transitTimeCutClassesDef)]

# remove all-0 rows after filtering
Asnicar_complete_no_NA_Constipation.tse <- Asnicar_complete_no_NA_Constipation.tse[rowSums(assay(Asnicar_complete_no_NA_Constipation.tse, "relabundance")) > 0,]


```

## Fit Asnicar Data (`asin(Hellinger)`)

```{r}
slow_vs_normal <- limmaFit_TSE(Asnicar_complete_no_NA_Constipation.tse, assay.type = "asin_hellinger", formula = ~ transitTimeCutClassesDef, coef = "transitTimeCutClassesDefSlow")

slow_vs_normal_prev5pct <- slow_vs_normal %>% 
  filter(N_not_zero/N_zero >= 0.05)

fast_vs_normal <- limmaFit_TSE(Asnicar_complete_no_NA_Constipation.tse, assay.type = "asin_hellinger", formula = ~ transitTimeCutClassesDef, coef = "transitTimeCutClassesDefFast")

fast_vs_normal_prev5pct <- fast_vs_normal%>% 
  filter(N_not_zero/N_zero >= 0.05)

inner_join(fast_vs_normal_prev5pct, slow_vs_normal_prev5pct, by = "SGB") %>% 
  ggplot(aes(x = Beta.x, y = Beta.y)) + 
  geom_point() + 
  geom_abline(intercept = 0, slope = 1, color = "red") + 
  geom_abline(intercept = 0, slope = -1, color = "red") + 
  stat_cor(method = "pearson") + 
  geom_smooth(method = "lm", formula = y ~ x) + 
  labs(
    title = "Effect Size scatterplot",
    subtitle = "baseline: Normal transit time", 
    x = "Fast Transit",
    y = "Slow Transit"
  )
```

Associations of fast and slow transit against normal transit return negatively
correlated associations, meaning a considerable portion of bugs that are 
associated in one direction in the slow transit are in the opposite direction in
the fast transit and vice versa.

# WallenZD_2022

```{r}
WallenZD_2022.tse <- sampleMetadata %>% 
  filter(study_name == "WallenZD_2022") %>% 
  returnSamples(data_type = "relative_abundance") %>% 
  pMD_enhance_MetaPhlAn()

# remove swab samples
WallenZD_2022_noSwabs.tse <- WallenZD_2022.tse[, WallenZD_2022.tse@colData$uncurated_collection_method != "swab"]

# prune resulting table
WallenZD_2022_noSwabs.tse <- WallenZD_2022_noSwabs.tse[rowSums(assay(WallenZD_2022_noSwabs.tse)) > 0,]

# calculate asin(sqrt(relabundance))
assay(WallenZD_2022_noSwabs.tse, "asin_hellinger") <- asin(sqrt(assay(WallenZD_2022_noSwabs.tse, "relabundance")))

table(WallenZD_2022_noSwabs.tse@colData$uncurated_constipation, WallenZD_2022_noSwabs.tse@colData$control, useNA = "ifany")
```

## Fit HEALTHY individuals in Wallen data (`asin(Hellinger)`)

```{r}
WallenZD_2022_noSwabs_noPD.tse <- WallenZD_2022_noSwabs.tse[, WallenZD_2022_noSwabs.tse@colData$control == 'Study Control' &  WallenZD_2022_noSwabs.tse@colData$uncurated_constipation != ""]

WallenZD_2022_noSwabs_noPD.tse <- WallenZD_2022_noSwabs_noPD.tse[rowSums(assay(WallenZD_2022_noSwabs_noPD.tse)) > 0, ]

# recode constipation
WallenZD_2022_noSwabs_noPD.tse@colData$uncurated_constipation <- factor(WallenZD_2022_noSwabs_noPD.tse@colData$uncurated_constipation, levels = c("N", "Y"))

wallen_fit_Controls_constipated_vs_not_constipated <- limmaFit_TSE(WallenZD_2022_noSwabs_noPD.tse, assay.type = "asin_hellinger", formula = ~ uncurated_constipation, coef = "uncurated_constipationY") 

wallen_fit_Controls_constipated_vs_not_constipated_5perc <- wallen_fit_Controls_constipated_vs_not_constipated %>% 
   filter(N_not_zero/N_zero >= 0.05)

arrange(wallen_fit_Controls_constipated_vs_not_constipated_5perc, adj.P.Value)
```
